### ISSUE TITLE
IFrame/Embeddable is blank on medium

### ISSUE THREAD

**Issue by mittalyashu** [dawnlabs/carbon#549]
**Describe the bug**

I have tried pasting the carbon url. But nothing is showing on the screen.

https://medium.com/@mittalyashu/omo-b71b2a7529e2

**To Reproduce**
1. Open Medium
2. Create new story
3. Try pasting any Carbon URL

**Expected behavior**

It should show the embed carbon on medium

**Screenshots**

![image](https://user-images.githubusercontent.com/29014463/46966625-de38c100-d0cb-11e8-830e-3de774500492.png)

**Info (please complete the following information):**
 - OS [e.g. macOS, Linux, Windows, iOS]: windows 10
 - Browser [e.g. chrome, safari]: chrome

**Comment by mfix22** [comment]
Can you copy paste your Carbon URL here? There are some limitations in Medium that we just can't get around 🙂

**Comment by mittalyashu** [comment]
Sure.

Here's the link: https://carbon.now.sh/?bg=rgba(191%2C220%2C238%2C1)&t=seti&wt=none&l=javascript&ds=true&dsyoff=20px&dsblur=73px&wc=true&wa=true&pv=48px&ph=32px&ln=false&fm=dm&fs=15px&lh=133%25&si=false&es=4x&wm=false&ts=true

**Comment by mfix22** [comment]
Here is the correct URL: https://carbon.now.sh/?bg=rgba(191%2C220%2C238%2C1)&t=seti&wt=none&l=javascript&ds=true&dsyoff=20px&dsblur=73px&wc=true&wa=true&pv=48px&ph=32px&ln=false&fm=dm&fs=15px&lh=133%25&si=false&es=4x&wm=false&ts=true 🙂

**Comment by mittalyashu** [comment]
But what is the difference in links. Both of them looks same to me.

**Comment by mfix22** [comment]
I think it was a matter of encoding. If you click the copy embed button everything should be correct, however there might be issues when copy pasting it from the Browser URL bar right now. 

In the console  run `encodeURI('<your url string>')` to if you see a failure (:

@mittalyashu would you be able to work on making sure the `window.location` URLs are always encoded properly?

**Comment by mittalyashu** [comment]
> In the console run encodeURI('<your url string>') to if you see a failure (:

It returned the same url when I run this command in console.

> @mittalyashu would you be able to work on making sure the window.location URLs are always encoded properly?

How? 😅

**Comment by VinGarcia** [comment]
Hello all, it seems this problem has returned. I only see a white clock when I embed from carbon even using the "Copy Embed" option and even when I use the correct link you posted:

> Here is the correct URL: https://carbon.now.sh/?bg=rgba(191%2C220%2C238%2C1)&t=seti&wt=none&l=javascript&ds=true&dsyoff=20px&dsblur=73px&wc=true&wa=true&pv=48px&ph=32px&ln=false&fm=dm&fs=15px&lh=133%25&si=false&es=4x&wm=false&ts=true :slightly_smiling_face: 

I also read on [a Medium Suport page](https://help.medium.com/hc/en-us/articles/214981378-Embeds) that:

> Embeds on Medium are handled through Embed.ly, a dedicated embedding service

So I wen't there and they say that they work with carbon, but it also failed on my tests:

- [Test with URL mentioned in this thread](https://embed.ly/code?url=https%3A%2F%2Fcarbon.now.sh%2F%3Fbg%3Drgba(191%252C220%252C238%252C1)%26t%3Dseti%26wt%3Dnone%26l%3Djavascript%26ds%3Dtrue%26dsyoff%3D20px%26dsblur%3D73px%26wc%3Dtrue%26wa%3Dtrue%26pv%3D48px%26ph%3D32px%26ln%3Dfalse%26fm%3Ddm%26fs%3D15px%26lh%3D133%2525%26si%3Dfalse%26es%3D4x%26wm%3Dfalse%26ts%3Dtrue
- [Test with a normal URL carbon.now url I just generated](https://embed.ly/code?url=https%3A%2F%2Fcarbon.now.sh%2F%3Fbg%3Drgba(171%252C184%252C195%252C0)%26t%3Done-light%26wt%3Dbw%26l%3Djavascript%26ds%3Dtrue%26dsyoff%3D20px%26dsblur%3D68px%26wc%3Dfalse%26wa%3Dtrue%26pv%3D0px%26ph%3D0px%26ln%3Dfalse%26fm%3DHack%26fs%3D14px%26lh%3D133%2525%26si%3Dfalse%26es%3D4x%26wm%3Dfalse)
- Test with a <iframe> from Copy Embed doesn't work as it expects a link.

I hope this is helpful.

Regards

**Comment by mfix22** [comment]
@VinGarcia 
> Here is the correct URL: https://carbon.now.sh/?bg=rgba(191%2C220%2C238%2C1)&t=seti&wt=none&l=javascript&ds=true&dsyoff=20px&dsblur=73px&wc=true&wa=true&pv=48px&ph=32px&ln=false&fm=dm&fs=15px&lh=133%25&si=false&es=4x&wm=false&ts=true 🙂

This URL: https://carbon.now.sh/?bg=rgba(191%252C220%252C238%252C1)&t=seti&wt=none&l=javascript&ds=true&dsyoff=20px&dsblur=73px&wc=true&wa=true&pv=48px&ph=32px&ln=false&fm=dm&fs=15px&lh=133%2525&si=false&es=4x&wm=false&ts=true works for me. There is a bug where you need to run `encodeURI(yourURLString)` again to make it work on Medium. I'll make an issue to fix that. 


As for Embed.ly working, that might be the same issue, but it might be a separate one that I need to investigate. Will open a new issue if fixing this doesn't fix the other.

**Comment by CHBaker** [comment]
Hey, I'm still getting this issue: this is the link I'm trying to paste.
https://carbon.now.sh/embed/?bg=rgba(74%252C121%252C226%252C1)&t=theme%253Ay9vkpjo7vj&wt=none&l=application%252Ftypescript&ds=false&dsyoff=20px&dsblur=68px&wc=true&wa=false&pv=16px&ph=100px&ln=false&fm=Hack&fs=13px&lh=149%2525&si=false&es=4x&wm=false&code=%25252F%25252F%252520We%252520get%252520a%252520very%252520simple%252520little%252520decorator%25250A%252540onChanges(%25250A%252520%252520%25252F%25252F%252520this%252520is%252520our%252520typed%252520callback%25252C%252520we%252520could%252520even%252520pass%252520it%252520as%252520a%252520method%25250A%252520%252520(oldVal%25253A%252520string%25252C%252520newVal%25253A%252520string)%252520%25253D%25253E%252520%25257B%25250A%252520%252520%252520%252520console.log('value%252520of%252520property%252520title%252520has%252520changed%252520%25255Cn')%25253B%25250A%252520%252520%252520%252520console.log('oldVal%25253A%252520'%25252C%252520oldVal%25252C%252520'%25255Cn%25255Cn%25255Cn')%25253B%25250A%252520%252520%252520%252520console.log('newVal%25253A%252520'%25252C%252520newVal)%25253B%25250A%252520%252520%25257D%25250A)%25250Ataco%252520%25253D%252520'High%252520Steaks%252520Beef%252520Bomb'%25253B

**Comment by samermurad** [comment]
Just in case someone comes here with the same frustration, I managed to get it working, always, by doing the following:

- It's important to not be logged in into carbon with github, so make sure you logout first
- Write the code that you intend to post (split it to a couple of snippets if it's too long, medium won't convert it otherwise)
-  Click on the clipboard icon and select "Plain URL", this will copy the URL.
- open up the dev console, and paste the following method:

```
function fixCarbonLinkForMedium(link) {
    const text = document.createElement('input')
    text.value = encodeURI(link.replace('sh/embed', 'sh').replace(/sh\/?/, 'sh/embed'))
    return text.value
}
```
- copy the the new logged URL, paste it in your medium article and click enter.

Hope this helps anyone with the same problem:D

**Comment by mfix22** [comment]
@samermurad I'm a little confused what this snippet does^. 

If you click the "Copy Medium.com" you will get the correct URL that does _not_ contain `/embed` right? You only need to do this if you are using the "Plain URL" which, like you mentioned, doesn't work in Medium since they expect a specific encoding.

**Comment by samermurad** [comment]
@mfilenko the JS snippet does both, the `encodeURI` fixes the encoding and the Regex replacement ensures that the /embed path is also there, pasting the outputted link in medium works, in edit mode, it will be displayed as an image, and will be selectable/scrollable when you publish.
That’s the way i got it working, “Copy Medium.com”, at least for me, pasted an image, and stayed an image after publishing.
So the above is what I needed to do, to make this work as a selectable, scrollable chunk of code

**Comment by tanusgit** [comment]
Still not working for me. pasted this link to medium and all it shows is the same url text, not even the image!

https://carbon.now.sh/?bg=rgba%2528171%252C+184%252C+195%252C+1%2529&t=seti&wt=none&l=text%252Fx-java&width=680&ds=true&dsyoff=20px&dsblur=68px&wc=false&wa=true&pv=6px&ph=9px&ln=false&fl=1&fm=Hack&fs=14px&lh=133%2525&si=false&es=2x&wm=false&code=public%252520static%252520int%252520max%2528int%252520s%2529%25257B%25250A%252520%252520%252520%252520int%252520max%252520%25253D%252520Integer.MIN_VALUE%25253B%25250A%252520%252520%252520%252520String%252520str%252520%25253D%252520Integer.toString%2528s%2529%25253B%25250A%252520%252520%252520%252520for%2528int%252520i%252520%25253D0%25253B%252520i%252520%25253C%252520str.length%2528%2529%25253B%252520i%25252B%25252B%2529%25257B%25250A%252520%252520%252520%252520%252520%252520%252520%252520char%252520c%252520%25253D%252520str.charAt%2528i%2529%25253B%25250A%252520%252520%252520%252520%252520%252520%252520%252520if%2528c%252520%25253D%25253D%252520%25275%2527%2529%25257B%25250A%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%25252F%25252Fmake%252520a%252520string%252520after%252520deleting%252520the%2525205%25252C%252520thus%252520we%252520make%252520many%252520strings%252520in%252520each%252520iteration%25250A%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%25252F%25252Fin%252520first%252520iteration%252520the%252520first%2525205%252520and%252520then%252520in%252520the%252520next%252520iteration%252520the%252520next%2525205%252520keeping%25250A%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%25252F%25252Fthe%252520previous%2525205%252520in%252520the%252520string%25250A%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520String%252520res%252520%25253D%252520str.substring%25280%25252C%252520i%2529%252520%25252B%252520str.substring%2528i%25252B1%2529%25253B%25250A%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520int%252520r%252520%25253D%252520Integer.parseInt%2528res%2529%25253B%25250A%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520if%2528r%252520%25253E%252520max%2529%25257B%25250A%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520max%252520%25253D%252520r%25253B%25250A%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%25257Delse%25257B%25250A%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520max%252520%25253D%252520max%25253B%25250A%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%25257D%25250A%252520%252520%252520%252520%252520%252520%252520%252520%25257D%25250A%252520%252520%252520%252520%25257D%25250A%252520%252520%252520%252520return%252520max%25253B%25250A%25257D

**Comment by mahci** [comment]
For me also the embed links not working properly. This is the link I get when I click on copy > medium.com:

https://carbon.now.sh/?bg=rgba%2528100%252C134%252C161%252C1%2529&t=theme%253A8ke3uw77yvm&wt=none&l=text%252Fx-java&width=680&ds=true&dsyoff=20px&dsblur=68px&wc=true&wa=true&pv=56px&ph=56px&ln=false&fl=1&fm=Hack&fs=14px&lh=133%2525&si=false&es=2x&wm=false&code=import%252520javax.sound.sampled.AudioSystem%25253B%25250Aimport%252520javax.sound.sampled.Clip%25253B%25250Aimport%252520javax.sound.sampled.LineUnavailableException%25253B%25250Aimport%252520javax.sound.sampled.UnsupportedAudioFileException%25253B%25250Aimport%252520java.io.File%25253B%25250Aimport%252520java.io.IOException%25253B%25250A%25250Apublic%252520class%252520SoundPlayer%252520%25257B%25250A%25250A%252520%252520%252520%252520private%252520static%252520SoundPlayer%252520instance%25253B%25250A%25250A%252520%252520%252520%252520private%252520Clip%252520mClip%25253B%25250A%25250A%252520%252520%252520%252520public%252520static%252520SoundPlayer%252520get%2528%2529%252520%25257B%25250A%252520%252520%252520%252520%252520%252520%252520%252520if%252520%2528instance%252520%25253D%25253D%252520null%2529%252520instance%252520%25253D%252520new%252520SoundPlayer%2528%2529%25253B%25250A%252520%252520%252520%252520%252520%252520%252520%252520return%252520instance%25253B%25250A%252520%252520%252520%252520%25257D%25250A%25250A%252520%252520%252520%252520private%252520SoundPlayer%2528%2529%252520%25257B%25250A%252520%252520%252520%252520%252520%252520%252520%252520final%252520File%252520soundFile%252520%25253D%252520new%252520File%2528%252522.%25252Fres%25252Fmy_sound.wav%252522%2529%25253B%25250A%25250A%252520%252520%252520%252520%252520%252520%252520%252520try%252520%25257B%25250A%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520mClip%252520%25253D%252520AudioSystem.getClip%2528%2529%25253B%25250A%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520mClip.open%2528AudioSystem.getAudioInputStream%2528soundFile%2529%2529%25253B%25250A%25250A%252520%252520%252520%252520%252520%252520%252520%252520%25257D%252520catch%252520%2528LineUnavailableException%252520%25257C%252520IOException%252520%25257C%252520UnsupportedAudioFileException%252520e%2529%252520%25257B%25250A%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520%252520throw%252520new%252520RuntimeException%2528e%2529%25253B%25250A%252520%252520%252520%252520%252520%252520%252520%252520%25257D%25250A%252520%252520%252520%252520%25257D%25250A%25250A%25257D

When pasted on Medium, it just shows this link, no 'embedding' happens.

**Comment by JacobBumgarner** [comment]
@tanusgit @mahci @mfix22 

I was experiencing a similar issue with my code snippets. It turns out that there isn't an issue on Carbon's end, but instead on Medium's end.

When the URL is too long, Medium won't embed it.

Instead, try pasting your URL into tinyurl.com and embed it that way.

https://user-images.githubusercontent.com/70919881/193323195-ba797910-3b3c-4a59-b2e2-2f2b1d315211.mov


I pulled my hair out WAY too long until I finally figured this one out. Figured I would share the solution with you.

As a brief note, if the code block from carbon is too long, Medium won't embed the link and will just instead show a link to the Carbon website. However, I'd imagine that most code snippets should be short enough to avoid that problem.
